src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  actions.forEach((action) => {
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      action.keysToRemove.forEach((key) => {
+        delete currentState[key];
+      });
+    } else if (action.type === 'clear') {
+      currentState = {};
+    }
+    stateHistory.push(currentState);
```

Comment:
You are pushing the reference to the same `currentState` object into the `stateHistory` array. This will result in all entries in the `stateHistory` array pointing to the same object, which is the final state. To fix this, push a copy of `currentState` instead to ensure that each entry in the array is a snapshot of the state at that point in time.